File: exercicios/python/Exercicio 04 - Relações entre Objetos (01_09)/livro.py
from editora import Editora
from autor import Autor
from capitulo import Capitulo

class Livro:
    def __init__(self, codigo: int, titulo: str, ano: int, editora: Editora, autor: Autor, numero_capitulo: int, titulo_capitulo: str):
        self.__codigo = codigo
        self.__titulo = titulo
        self.__ano = ano
        self.__editora = editora
        self.__numero_capitulo = numero_capitulo
        self.__titulo_capitulo = titulo_capitulo
        self.__autores = [autor]
        self.__capitulos = [Capitulo(numero_capitulo,titulo_capitulo)]

    @property
    def codigo(self):
        return self.__codigo

    @codigo.setter
    def codigo(self, codigo):
        self.__codigo = codigo
        return self.__codigo
    
    @property
    def titulo(self):
        return self.__titulo
    
    @titulo.setter
    def titulo(self, titulo):
        self.__titulo = titulo

    @property
    def ano(self):
        return self.__ano
    
    @ano.setter
    def ano(self,ano):
        self.__ano = ano
    
    @property
    def editora(self):
        return self.__editora
    
    @editora.setter
    def editora(self, editora):
        self.__editora = editora

    @property
    def autores(self):
        return self.__autores
    
    @property
    def capitulos(self):
        return self.__capitulos

    def incluir_autor(self, autor: Autor):
        if isinstance(autor, Autor) and autor not in self.__autores:
            self.__autores.append(autor)

    def excluir_autor(self, autor: Autor):
        self.__autores.remove(autor)

    def incluir_capitulo(self, numero: int, titulo: str):
        capitulo = self.find_capitulo_by_titulo(titulo)
        if isinstance(numero, int) and isinstance(titulo,str) and capitulo == None:
            self.__capitulos.append(Capitulo(numero,titulo))

    def excluir_capitulo(self, titulo: str):
        capitulo = self.find_capitulo_by_titulo(titulo)
        if capitulo != None:
            self.__capitulos.remove(capitulo)

    def find_capitulo_by_titulo(self, titulo: str):
        for capitulo in self.__capitulos:
            if capitulo.titulo == titulo:
                return capitulo
        return None
    

File: exercicios/java/dsoex04/src/main/java/org/example/Livro.java
package org.example;

import java.util.ArrayList;
import java.util.List;

public class Livro {
    private int codigo;
    private String titulo;
    private int ano;
    private Editora editora;
    private List<Autor> autores;
    private List<Capitulo> capitulos;

    public Livro(int codigo, String titulo, int ano, Editora editora, Autor autor, int numeroCapitulo, String tituloCapitulo) {
        this.codigo = codigo;
        this.titulo = titulo;
        this.ano = ano;
        this.editora = editora;
        this.autores = new ArrayList<>();
        this.autores.add(autor);
        this.capitulos = new ArrayList<>();
        this.capitulos.add(new Capitulo(numeroCapitulo, tituloCapitulo));
    }

    public int getCodigo() {
        return codigo;
    }
    public void setCodigo(int codigo) {
        this.codigo = codigo;
    }

    public String getTitulo() {
        return titulo;
    }
    public void setTitulo(String titulo) {
        this.titulo = titulo;
    }
    public int getAno() {
        return ano;
    }
    public void setAno(int ano) {
        this.ano = ano;
    }
    public Editora getEditora() {
        return editora;
    }
    public void setEditora(Editora editora) {
        this.editora = editora;
    }
    public List<Autor> getAutores() {
        return autores;
    }
    public List<Capitulo> getCapitulos() {
        return capitulos;
    }
    public void incluirAutor(Autor autor) {
        if (autor != null && !this.autores.contains(autor)) {
            this.autores.add(autor);
        }
    }
    public void incluirCapitulo(int numero, String titulo) {
        Capitulo capitulo = findCapituloByTitulo(titulo);
        if (capitulo == null) {
            this.capitulos.add(new Capitulo(numero,titulo));}
    }
    public Capitulo findCapituloByTitulo(String titulo){
        for (Capitulo capitulo : this.capitulos){
            if (capitulo.getTitulo().equals(titulo)){
                return capitulo;
            }
        }
        return null;
    }
    public void excluirAutor(Autor autor) {
        if (autor != null && this.autores.contains(autor)) {
            this.autores.remove(autor);
        }
    }
    public void excluirCapitulo(String titulo) {
        Capitulo capitulo = findCapituloByTitulo(titulo);
        if (capitulo != null && this.capitulos.contains(capitulo)) {
            this.capitulos.remove(capitulo);
        }
    }
}